### DISCUSSION TITLE
JS外语言，如Python/CPP的无法调试

### DISCUSSION THREAD

**Discussion by Hitooooo** [opensumi/core#2831] [Q&A]
使用CodeRunner插件是可以执行Python/CPP代码的，但是使用vscode插件python(v2020.3.71113)和C/C++(v1.5.1)均无法调试。
python报错：
```command not found: cmd```.
![image](https://github.com/opensumi/core/assets/19512047/71fd6c68-a131-487b-a70d-900e1543cca2)

cpp报错：
![image](https://github.com/opensumi/core/assets/19512047/de4733dc-72cf-4792-ad47-c0ac3aa5541f)

执行环境：WSL2
版本：[官方cloud ide](https://github.com/opensumi/ide-startup)

launch.json
```json
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
    {
      "name": "C/C++: g++ build and debug active file",
      "type": "cppdbg",
      "request": "launch",
      "program": "${fileDirname}/${fileBasenameNoExtension}",
      "args": [],
      "stopAtEntry": false,
      "cwd": "${workspaceFolder}",
      "environment": [],
      "externalConsole": false,
      "MIMode": "gdb",
      "miDebuggerPath": "/usr/bin/gdb",
      "setupCommands": [
        {
          "description": "Enable pretty-printing for gdb",
          "text": "-enable-pretty-printing",
          "ignoreFailures": true
        }
      ],
      "preLaunchTask": "C/C++: g++ build active file"
    },
    {
      "name": "Python: Current File",
      "type": "python",
      "request": "launch",
      "program": "${file}",
      "console": "integratedTerminal"
    }
  ]
}
```

不明白Linux环境为什么会执行windows平台cmd命令，这些插件能升级至vscode最新版本吗？

**Comment by erha19** [comment]
@Hitooooo 部分插件针对系统环境是需要区分版本的，你可以从 https://open-vsx.org/ 或 https://marketplace.visualstudio.com/vscode 中找到对应系统环境下的插件进行部署。

> 不明白Linux环境为什么会执行windows平台cmd命令，这些插件能升级至vscode最新版本吗？

这个实际上取决于插件实现，与框架关系不大，以 C++ 为例，你可以试一下升级插件或寻找其他可用插件，如 https://open-vsx.org/extension/franneck94/c-cpp-runner ，而针对 python 插件，由于高版本依赖了 notebook API 并不支持，可能会有部分使用问题。

**Reply by Hitooooo** [reply]
ok~明白。 那部署Linux版本的插件，需要手动修改scripts/download.js实现吗？ 比如期望安装https://github.com/microsoft/vscode-cpptools/releases/download/v1.16.3/cpptools-linux.vsix插件，利用download.js解压至extentions目录下即可？
有没有什么方便的方式？

**Comment by Hitooooo** [comment]
@erha19 已经部署了https://github.com/microsoft/vscode-cpptools/releases/download/v1.16.3/cpptools-linux.vsix 问题仍无法解决，sumi兼容这个框架了吗？

**Comment by erha19** [comment]
@Hitooooo 插件相关适配，目前 OpenSumi 只提供了到 VS Code 1.69.0 版本标准 API，同时如 notebook，timeline 等 API 是做了空实现，没有实际功能，需要自行评估插件是否可用。

同时，对于插件不能使用的情况，由于受测试环境限制，我们建议开发者自行通过如下方式进行排查:

1. 查看控制台日志信息，看是否有内核版本上的要求，如：https://github.com/opensumi/core/discussions/2011#discussioncomment-4223271 ， 这种情况需要修改内核版本以适配插件，参考格式如：https://github.com/opensumi/ide-startup/blob/bd9327994db462e2d46c214347cd293178e3cc3b/src/extension/ext-host.ts
2. 调试插件进程，看是否存在方法未实现的报错或其他信息，建议直接运行 opensumi/core 项目，调试文档见：https://opensumi.com/zh/docs/develop/how-to-contribute#%E6%8F%92%E4%BB%B6%E8%BF%9B%E7%A8%8B

**Comment by Hitooooo** [comment]
在CloudIDE场景中，后台是Linux但在Windows的浏览器中打开，那么platform.ts中获取到的```_isWindows```属性为true，导致terminal.ts中```prepareCommand()```拼接的命令行加上了 ```cmd \C```。
根据业务需求修改即可。

**Reply by erha19** [reply]
相关判断代码可以看一下是哪段？框架目前有提供 BackgroundOS 来解决这个问题，这个可能跟插件的实现有一定关联

**Reply by Hitooooo** [reply]
这应该还没到插件的部分吧，看起来是控制台执行命令逻辑。具体BackgroundOS有相关文档或示例代码吗？

**Comment by Hitooooo** [comment]
![image](https://github.com/opensumi/core/assets/19512047/236bf4f6-ff87-4dd2-9176-dc1682b746c8)

获取的是浏览器运行所在系统的版本。

![image](https://github.com/opensumi/core/assets/19512047/04e00f4a-f65d-4669-84c8-2493651d2a27)

最终生成的终端执行命令。

**Comment by erha19** [comment]
@Hitooooo 了解了，这里就是环境判断导致的问题，应该使用 backendOS 来进行判断，使用示例如：
```ts
...
@Autowired(IApplicationService)
private readonly appService: IApplicationService;
...

if ((await this.appService.backendOS) === OperatingSystem.Windows) {
 // Windows 环境判断
}
...
```

见：https://github.com/opensumi/core/blob/5bea21827177d7f2db7098edc5517c63b86dcbcc/packages/file-tree-next/src/browser/file-tree-contribution.ts#L746

这里相关代码可能得从 common 移动到 browser 中进行判断，其中 common 在框架内是用于存放 browser 及 node 的公共部分逻辑，不能在这里引入 browser 相关内容。
